Add COND speculator to DynamoRIO backend

AlviseDeFaveri opened the pull request:
This PR adds the COND speculator to Revizor's DR backend and fixes some related bugs that have surfaced by executing non-architectural paths, mainly:
- handling AVX instructions
- handling speculative exceptions
- handling speculative syscalls
- handling speculative TSX instructions
- handling speculative XSAVE/XRESTORE instructions
- fixing broken rollback logic
- handling cases in which rollbacks can fail, namely
    - stores to invalid addresses
    - stores to non-writable memory
    - stores to memory incorrectly marked by DynamoRIO as executable (see comment in `SpeculatorABC::rollback()`)

OleksiiOleksenko commented on rvzr/model_dynamorio/backend/include/speculator_abc.hpp:
```diff
@@ -98,8 +98,8 @@ class SpeculatorABC
     /// @param size The size of the memory access
-    /// @return void
-    virtual void handle_mem_access(bool is_write, void *address, uint64_t size);
+    /// @return false if the memory access cannot be rolled-back
```
The phrasing is a bit confusing. Maybe "false if the memory access is invalid and is going to produce an exception"?

OleksiiOleksenko commented on rvzr/model_dynamorio/backend/include/util.hpp:
```diff
@@ -33,3 +35,21 @@ void reserve_register_checked(void *drcontext, instrlist_t *ilist, instr_t *wher
+/// @param instr The instruction to check
+/// @param mc
+/// @param dc
+bool is_illegal_jump(instr_obs_t instr, dr_mcontext_t *mc, void *dc);
```
It would be helpful to leave a note in the docstring mentioning that this function is unused, and why.

OleksiiOleksenko commented on rvzr/model_dynamorio/backend/speculator_abc.cpp:
```diff
@@ -108,9 +108,35 @@ pc_t SpeculatorABC::rollback(dr_mcontext_t *mc)
+        //       2. Rollback after faulty indirect call/ret: if a data ddress is used as an
+        //       (invalid) call or ret target, DynamoRIO will wrongly mark the address as belonging
+        //       to the code cache, and therefore prohibit writes.
+        if (not success and not(store_log.size() > last_committed)) {
```
This starts to resemble the Hadouken Pattern. Perhaps move this into a separate private method?

OleksiiOleksenko commented on rvzr/model_dynamorio/backend/speculator_abc.cpp:
```diff
@@ -146,17 +147,44 @@ pc_t SpeculatorABC::handle_instruction(instr_obs_t instr, dr_mcontext_t *mc, voi
-        // so we just accept that horrible things may happen. Oh well.
-        const uint64_t val = *(uint64_t *)address;
-        store_log.push_back({.addr = (uint64_t)address, .val = val, .nesting_level = nesting});
+        auto cur_address = (uint64_t)address;
```
The code will be slightly easier to read if you rewrite it to `if not is_write -> return`

AlviseDeFaveri commented:
Addressed comments + made some changes after observing other edge cases.

All new changes are for the commit about unhandled rollback stores. TL;DR
- Forcing writes is not safe and will cause DynamoRIO to abort in some cases. The sound thing to do is to check for illegal jumps and rollback before executing them
- Fixed `is_illegal_jump` to handle more cases and made it more readable
- Encapsulated all logic for committed/in-flight `StoreLog` object that has the same interface as `std::vector` + some logic to flush uncommitted entries
- Added a `handle_exception` function to `SpeculatorABC` to flush all uncommitted entries from the `store_log` upon exceptions
- This way, there is no need for special-casing when rolling back stores: all of these edge-cases arecare of before the rollback starts
